src/lib/shim/shadow_sem.c

```diff
@@ -0,0 +1,155 @@
+#include "lib/shim/shadow_sem.h"
+
+#include "lib/logger/logger.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <linux/futex.h>
+#include <semaphore.h>
+#include <stdalign.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+static inline int* _shadow_sem_futex_addr(const shadow_sem_t* s) {
```

Comment:
OK, last nits :)

1. I think we should drop the `const` on the parameter here and at Line 23 --- a `FUTEX_WAKE` operation will modify the futex word, right?

2. Based on current usage (i.e. immediately passing the returned address to `syscall`), I don't think this is necessary, but I think the return should `_Atomic` or at least `volatile`(?)

Reply:
> I think we should drop the const on the parameter here and at Line 23 --- a FUTEX_WAKE operation will modify the futex word, right?

The FUTEX_WAKE doesn't modify the futex word; the waker modifies the word (with an atomic op) before doing the FUTEX_WAKE operation. Dropped the const anyway; see below.

> Based on current usage (i.e. immediately passing the returned address to syscall), I don't think this is necessary, but I think the return should _Atomic or at least volatile(?)

I changed this to keep the current pointer type instead of casting it. I ended up dropping the `const` for that to work (otherwise we need to cast away the `const` in the wake helper).